tests/shell/tests.py

```diff
@@ -112,13 +130,151 @@ def test_python(self):
+            self.assertNotIn(article_shell, namespace.values())
+
+    @override_settings(
+        INSTALLED_APPS=["shell", "django.contrib.auth", "django.contrib.contenttypes"]
+    )
+    def test_overridden_get_namespace(self):
+        class Command(shell.Command):
+            def get_namespace(self):
+                from django.urls.base import resolve, reverse
+
+                return {
+                    **super().get_namespace(),
+                    "resolve": resolve,
+                    "reverse": reverse,
+                }
+
+        cmd = Command()
+        namespace = cmd.get_namespace()
+
+        self.assertEqual(
+            namespace,
+            {
+                "resolve": resolve,
+                "reverse": reverse,
+                "Marker": shell_models.Marker,
+                "Phone": shell_models.Phone,
+                "ContentType": ContentType,
+                "Group": Group,
+                "Permission": Permission,
+                "User": User,
+                "auth_models": auth_model_module,
+                "contenttypes_models": contenttypes_model_module,
+                "shell_models": shell_models,
+            },
+        )
+
+    @override_settings(
+        INSTALLED_APPS=["shell", "django.contrib.auth", "django.contrib.contenttypes"]
+    )
+    def test_message_with_stdout(self):
+        with captured_stdout() as stdout:
+            cmd = shell.Command()
+            cmd.get_and_report_namespace(verbosity=1)
+        self.assertEqual(stdout.getvalue().strip(), "9 objects imported automatically")
+
+    @override_settings(
+        INSTALLED_APPS=["shell", "django.contrib.auth", "django.contrib.contenttypes"]
+    )
+    def test_message_with_no_stdout(self):
+        with captured_stdout() as stdout:
+            cmd = shell.Command()
+            cmd.get_and_report_namespace(verbosity=0)
+        self.assertEqual(stdout.getvalue().strip(), "")
+
+    @override_settings(INSTALLED_APPS=["shell", "django.contrib.contenttypes"])
+    def test_message_with_stdout_listing_objects(self):
+        class Command(shell.Command):
+            def get_namespace(self):
+                class MyClass:
+                    pass
+
+                return {
+                    **super().get_namespace(),
+                    "MyClass": MyClass,
+                }
+
+        with captured_stdout() as stdout:
+            cmd = Command()
+            cmd.get_and_report_namespace(verbosity=2)
+
+        self.assertEqual(
+            stdout.getvalue().strip(),
+            "from django.contrib.contenttypes.models import ContentType\n"
+            "from shell.models import Phone, Marker",
+        )
+
+    @override_settings(
+        INSTALLED_APPS=[
+            "shell",
+            "django.contrib.auth",
+            "django.contrib.messages",
+            "django.contrib.staticfiles",
+        ]
+    )
+    def test_shell_has_not_empty_models_objects(self):
```

Comment:
Slightly better naming suggestion. 😄 
```suggestion
    def test_shell_has_no_empty_models_objects(self):
```